cellpack/autopack/Compartment.py

```diff
@@ -1299,42 +1307,46 @@ def BuildGrid_trimesh(self, env, grdPos, vSurfaceArea, srfPts, idarray, mesh_sto
         ).hollow()
         self.log.info("VOXELIZED MESH")
         # the main loop
-        tree = spatial.cKDTree(grdPos, leafsize=10)
-        points_in_encap_sphere = tree.query_ball_point(
-            self.center,
-            self.encapsulating_radius + env.grid.gridSpacing * 2,
-            return_sorted=True,
+        tree = spatial.cKDTree(master_grid_positions, leafsize=10)
+        points_in_encap_sphere = numpy.array(
+            tree.query_ball_point(
+                self.center,
+                self.encapsulating_radius + env.grid.gridSpacing * 2,
+                return_sorted=True,
+            )
         )
         self.log.info(f"GOT POINTS IN SPHERE {len(points_in_encap_sphere)}")
-        for ptInd in points_in_encap_sphere:
-            coord = [
-                grdPos.item((ptInd, 0)),
-                grdPos.item((ptInd, 1)),
-                grdPos.item((ptInd, 2)),
-            ]
-            if idarray[ptInd] > 0:
-                continue
-            if trimesh_grid_surface.is_filled(coord):
-                idarray.itemset(ptInd, number)
-            elif mesh_store.contains_point(self.gname, coord):
-                insidePoints.append(ptInd)
-                idarray.itemset(ptInd, -number)
+
+        point_compartment_ids = compartment_ids[points_in_encap_sphere]
+        point_ids_to_assign = points_in_encap_sphere[point_compartment_ids == 0]
+        point_positions = master_grid_positions[point_ids_to_assign]
+
+        # check surface points
+        points_on_surface = trimesh_grid_surface.is_filled(point_positions)
+        compartment_ids[point_ids_to_assign[points_on_surface]] = number
+
+        # check inside points
+        points_in_mesh = mesh_store.contains_points_mesh(self.gname, point_positions)
```

Comment:
The logic of the changes you made to check for inside points and set the compartment id is so clear!